File: edexOsgi/com.raytheon.uf.common.dataquery/src/com/raytheon/uf/common/dataquery/requests/RequestConstraint.java
/**
 * This software was developed and / or modified by Raytheon Company,
 * pursuant to Contract DG133W-05-CQ-1067 with the US Government.
 * 
 * U.S. EXPORT CONTROLLED TECHNICAL DATA
 * This software product contains export-restricted data whose
 * export/transfer/disclosure is restricted by U.S. law. Dissemination
 * to non-U.S. persons whether in the United States or abroad requires
 * an export license or other authorization.
 * 
 * Contractor Name:        Raytheon Company
 * Contractor Address:     6825 Pine Street, Suite 340
 *                         Mail Stop B8
 *                         Omaha, NE 68106
 *                         402.291.0100
 * 
 * See the AWIPS II Master Rights File ("Master Rights File.pdf") for
 * further licensing information.
 **/
package com.raytheon.uf.common.dataquery.requests;

import java.lang.reflect.Array;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collection;
import java.util.Date;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlType;

import com.raytheon.uf.common.serialization.annotations.DynamicSerialize;
import com.raytheon.uf.common.serialization.annotations.DynamicSerializeElement;
import com.raytheon.uf.common.time.util.TimeUtil;

/**
 * Used in requests to limit the type of data returned. Similar to an sql WHERE
 * clause, it consists of a type(operator) and a value. When a request is made
 * fields will be compared to the constraint value with the specified type to
 * determine what data to return.
 * 
 * <pre>
 * 
 * SOFTWARE HISTORY
 *   
 * Date          Ticket#  Engineer    Description
 * ------------- -------- ----------- -----------------------------------------
 * Aug 21, 2007           chammack    Initial Creation.
 * May 27, 2009  2408     jsanchez    Cast value to String.
 * Sep 28, 2009  3099     bsteffen    Fixed constraintCompare to convert all
 *                                    non-numeric objects to String
 * Nov 05, 2009  3553     rjpeter     Added isNull capability.
 * Jul 09, 2013  1869     bsteffen    Format Calendar when making Constraint
 *                                    Mapping.
 * Dec 18, 2013  2579     bsteffen    Remove ISerializableObject
 * 
 * 
 * </pre>
 * 
 * @author chammack
 * @version 1
 */
@XmlAccessorType(XmlAccessType.NONE)
@XmlRootElement(name = "requestConstraint")
@XmlType(name = "requestConstraint")
@DynamicSerialize
public class RequestConstraint implements Cloneable {

    public static final RequestConstraint WILDCARD;
    static {
        WILDCARD = new RequestConstraint();
        WILDCARD.setConstraintType(ConstraintType.LIKE);
        WILDCARD.setConstraintValue("%");
    }

    private static final Pattern BETWEEN_PATTERN = Pattern.compile("--");

    private static final float EQUALITY_TOLERANCE = 0.0001f;

    public enum ConstraintType {
        EQUALS("="), NOT_EQUALS("!="), GREATER_THAN(">"), GREATER_THAN_EQUALS(
                ">="), LESS_THAN("<"), LESS_THAN_EQUALS("<="), BETWEEN(
                "between"), IN("in"), LIKE("like"), ILIKE("ilike"), ISNULL(
                "isnull"), ISNOTNULL("isnotnull");

        private String operand;

        private ConstraintType(String operand) {
            this.operand = operand;
        }

        public String getOperand() {
            return this.operand;
        }
    };

    private static EnumSet<ConstraintType> mergableTypesEqualIn = EnumSet.of(
            ConstraintType.EQUALS, ConstraintType.IN);

    @XmlAttribute
    @DynamicSerializeElement
    protected ConstraintType constraintType;

    @XmlAttribute
    @DynamicSerializeElement
    protected String constraintValue;

    protected transient Map<Class<?>, Object> asMap = new HashMap<Class<?>, Object>(
            2);

    /**
     * Constructor
     */
    public RequestConstraint() {
        this(ConstraintType.EQUALS);
    }

    public RequestConstraint(ConstraintType constraintType) {
        this(null, constraintType);
    }

    /**
     * Convenience constructor to construct an equals constraint
     * 
     * @param value
     *            the value to constrain on
     */
    public RequestConstraint(String value) {
        this(value, ConstraintType.EQUALS);
    }

    /**
     * Constructor for a specified constraint type and value
     * 
     * @param value
     * @param type
     */
    public RequestConstraint(String value, ConstraintType type) {
        this.constraintValue = value;
        this.constraintType = type;
    }

    /**
     * Converts inConstraints into String[] and calls
     * {@link RequestConstraint#RequestConstraint(String[])}
     * 
     * @param inConstraints
     */
    public RequestConstraint(Collection<String> inConstraints) {
        this(inConstraints.toArray(new String[0]));
    }

    /**
     * Creates a {@link RequestConstraint} with {@link ConstraintType#IN} with
     * inConstraints set as the {@link #setConstraintValueList(String[])} if
     * inConstraints size == 1 then {@link ConstraintType#EQUALS} will be used
     * instead
     * 
     * @param inConstraints
     */
    public RequestConstraint(String[] inConstraints) {
        this.constraintType = ConstraintType.IN;
        if (inConstraints.length == 1) {
            this.constraintType = ConstraintType.EQUALS;
            this.constraintValue = inConstraints[0];
        } else {
            setConstraintValueList(inConstraints);
        }
    }

    /**
     * Creates a {@link RequestConstraint} with {@link ConstraintType#BETWEEN}
     * 
     * @param low
     * @param high
     */
    public RequestConstraint(String low, String high) {
        this(ConstraintType.BETWEEN);
        setBetweenValueList(new String[] { low, high });
    }

    @Override
    public RequestConstraint clone() {
        return new RequestConstraint(constraintValue, constraintType);
    }

    public boolean merge(RequestConstraint that) {
        boolean successful = false;
        if (that != null && mergableTypesEqualIn.contains(this.constraintType)
                && mergableTypesEqualIn.contains(that.constraintType)) {
            if (this.constraintType.equals(ConstraintType.EQUALS)) {
                // upgrade to IN list
                this.constraintType = ConstraintType.IN;
            }

            this.addToConstraintValueList(that.constraintValue);
            successful = true;
        }
        return successful;
    }

    /**
     * @return the constraintType
     */
    public ConstraintType getConstraintType() {
        return constraintType;
    }

    /**
     * @param constraintType
     *            the constraintType to set
     */
    public void setConstraintType(ConstraintType constraintType) {
        this.constraintType = constraintType;
    }

    /**
     * @return the constraintValue
     */
    public String getConstraintValue() {
        return constraintValue;
    }

    /**
     * @param constraintValue
     *            the constraintValue to set
     */
    public void setConstraintValue(String constraintValue) {
        asMap.clear();
        this.constraintValue = constraintValue;
    }

    /**
     * 
     * @param constraintValues
     *            the constraintValues to set
     */
    public void setConstraintValueList(String[] constraintValues) {
        setConstraintValueList(Arrays.asList(constraintValues));
    }

    /**
     * Set a list of possible value for a request constraint, used for IN
     * ConstraintType.
     * 
     * @param constraintValues
     */
    public void setConstraintValueList(Collection<String> constraintValues) {
        StringBuffer sb = new StringBuffer(constraintValues.size() * 16);
        boolean first = true;
        for (String v : constraintValues) {
            if (!first) {
                sb.append(",");
            } else {
                first = false;
            }

            sb.append(v);
        }
        this.setConstraintValue(sb.toString());
    }

    public void addToConstraintValueList(String constraintValue) {
        if (this.constraintValue == null) {
            this.setConstraintValue(constraintValue);
        } else {
            this.setConstraintValue(this.constraintValue + ","
                    + constraintValue);
        }

    }

    /**
     * 
     * @param constraintValues
     *            the constraintValues to set
     */
    public void setBetweenValueList(String[] constraintValues) {
        StringBuffer sb = new StringBuffer();
        boolean first = true;
        for (String v : constraintValues) {
            if (!first) {
                sb.append("--");
            } else {
                first = false;
            }

            sb.append(v);
        }
        this.setConstraintValue(sb.toString());
    }

    /**
     * Evaluate whether a value satisfies a constraint
     * 
     * @param value
     *            the value
     * @return true if the value is satisfied by the constraint
     */
    public boolean evaluate(Object value) {
        if (this == WILDCARD) {
            return true;
        }

        if (constraintType == ConstraintType.ISNULL) {
            return value == null || "null".equals(value);
        } else if (constraintType == ConstraintType.ISNOTNULL) {
            return value != null && "null".equals(value) == false;
        }

        if (value == null) {
            return false;
        } else if (constraintType == ConstraintType.EQUALS) {
            return constraintCompare(value);
        } else if (constraintType == ConstraintType.NOT_EQUALS) {
            return (!constraintCompare(value));
        } else if (constraintType == ConstraintType.IN) {
            String[] list = constraintValue.split(",\\s?");
            if (value instanceof Number) {
                Double[] doubles = (Double[]) asMap.get(Double[].class);
                if (doubles == null) {
                    doubles = new Double[list.length];
                    for (int i = 0; i < list.length; i++) {
                        try {
                            doubles[i] = Double.valueOf(list[i]);
                        } catch (NumberFormatException e) {
                            continue;
                        }
                    }
                    asMap.put(Double[].class, doubles);
                }
                for (Double d : doubles) {
                    if (d == null) {
                        continue;
                    }
                    if (Math.abs(d.doubleValue()
                            - ((Number) value).doubleValue()) < EQUALITY_TOLERANCE) {
                        return true;
                    }
                }
                return false;
            } else {
                Arrays.sort(list);
                String searchValue = String.valueOf(value);
                if (Arrays.binarySearch(list, searchValue) > -1) {
                    return true;
                } else {
                    return false;
                }
            }
        } else if (constraintType == ConstraintType.LIKE) {
            String regex = constraintValue.replace("%", ".*");
            if (value.toString().matches(regex)) {
                return true;
            }
            return false;
        }

        if (value instanceof Date) {
            Date valueDate = (Date) value;

            if (constraintType == ConstraintType.BETWEEN) {
                String[] list = BETWEEN_PATTERN.split(constraintValue);
                if (list.length != 2) {
                    throw new IllegalArgumentException(
                            "Invalid between constraint.");
                }

                try {
                    DateFormat df = new SimpleDateFormat();
                    Date first = df.parse(list[0]);
                    Date last = df.parse(list[0]);
                    return (valueDate.equals(first) || valueDate.equals(last) || (valueDate
                            .after(first) && valueDate.before(last)));
                } catch (ParseException e) {
                    throw new IllegalArgumentException(
                            "Constraint does not appear to be a date");
                }

            }

            Date constraintValueDate;
            try {
                DateFormat df = new SimpleDateFormat();
                constraintValueDate = df.parse(constraintValue);
            } catch (ParseException e) {
                throw new IllegalArgumentException(
                        "Constraint does not appear to be a date");
            }

            if (constraintType == ConstraintType.GREATER_THAN) {
                return ((Date) value).after(constraintValueDate);
            } else if (constraintType == ConstraintType.GREATER_THAN_EQUALS) {
                return ((Date) value).after(constraintValueDate)
                        || constraintValue.equals(value);
            } else if (constraintType == ConstraintType.LESS_THAN) {
                return ((Date) value).before(constraintValueDate);
            } else if (constraintType == ConstraintType.LESS_THAN_EQUALS) {
                return ((Date) value).before(constraintValueDate)
                        || constraintValue.equals(value);
            }
        } else if (value instanceof Number) {
            double valueDouble = ((Number) value).doubleValue();

            if (constraintType == ConstraintType.BETWEEN) {
                String[] list = BETWEEN_PATTERN.split(constraintValue);
                if (list.length != 2) {
                    throw new IllegalArgumentException(
                            "Invalid between constraint.");
                }
                double lower = Double.valueOf(list[0]);
                double upper = Double.valueOf(list[1]);
                return (valueDouble >= lower && valueDouble <= upper);

            }

            double constraintValueDouble;

            try {
                constraintValueDouble = Double.parseDouble(constraintValue);
            } catch (NumberFormatException e) {
                throw new IllegalArgumentException(
                        "Constraint does not appear to be a number");
            }

            if (constraintType == ConstraintType.GREATER_THAN) {
                return (valueDouble > constraintValueDouble);
            } else if (constraintType == ConstraintType.GREATER_THAN_EQUALS) {
                return (valueDouble >= constraintValueDouble);
            } else if (constraintType == ConstraintType.LESS_THAN) {
                return (valueDouble < constraintValueDouble);
            } else if (constraintType == ConstraintType.LESS_THAN_EQUALS) {
                return (valueDouble <= constraintValueDouble);
            }

        }

        return false;
    }

    /**
     * @param value
     */
    private boolean constraintCompare(Object value) {
        if (value instanceof Number) {
            try {
                Double d = (Double) asMap.get(Double.class);
                if (d == null) {
                    d = Double.valueOf(constraintValue);
                    asMap.put(Double.class, d);
                }
                if (Math.abs(d.doubleValue() - ((Number) value).doubleValue()) < EQUALITY_TOLERANCE) {
                    return true;
                }
            } catch (NumberFormatException e) {
                return false;
            }
        } else {
            if (value != null) {
                return constraintValue.equals(value.toString());
            }
        }
        return false;
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#hashCode()
     */
    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result
                + ((constraintType == null) ? 0 : constraintType.hashCode());
        result = prime * result
                + ((constraintValue == null) ? 0 : constraintValue.hashCode());
        return result;
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#equals(java.lang.Object)
     */
    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final RequestConstraint other = (RequestConstraint) obj;
        if (constraintType == null) {
            if (other.constraintType != null) {
                return false;
            }
        } else if (!constraintType.equals(other.constraintType)) {
            return false;
        }
        if (constraintValue == null) {
            if (other.constraintValue != null) {
                return false;
            }
        } else if (!constraintValue.equals(other.constraintValue)) {
            return false;
        }
        return true;
    }

    /*
     * (non-Javadoc)
     * 
     * @see java.lang.Object#toString()
     */
    @Override
    public String toString() {
        return getConstraintType().name() + " " + getConstraintValue();
    }

    /**
     * Creates a constraint map based on an object field mapping.
     * {@link ConstraintType} will be to {@link ConstraintType#EQUALS} with
     * {@link Object#toString()} being called on the field value for values that
     * are not null. If field value is a {@link Collection} or an array type,
     * {@link ConstraintType#IN} will be used as the constraint type and each
     * field will have {@link Object#toString()} called on it. If field value is
     * null, {@link ConstraintType#ISNULL} will be used instead as the
     * {@link ConstraintType}
     * 
     * @param fieldMapping
     * @return
     */
    public static Map<String, RequestConstraint> toConstraintMapping(
            Map<String, Object> fieldMapping) {
        return toConstraintMapping(fieldMapping, true);
    }

    /**
     * Same functionality as {@link #toConstraintMapping(Map)} except null
     * values are not included in the resultant constraint map
     * 
     * @param fieldMapping
     * @return
     */
    public static Map<String, RequestConstraint> toConstraintMappingExcludeNull(
            Map<String, Object> fieldMapping) {
        return toConstraintMapping(fieldMapping, false);
    }

    private static Map<String, RequestConstraint> toConstraintMapping(
            Map<String, Object> fieldMapping, boolean includeNulls) {
        Map<String, RequestConstraint> constraints = new HashMap<String, RequestConstraint>();
        for (String key : fieldMapping.keySet()) {
            Object value = fieldMapping.get(key);
            ConstraintType constraintType = ConstraintType.EQUALS;
            String constraintValue = null;
            if (value == null) {
                constraintType = ConstraintType.ISNULL;
            } else if (value instanceof Calendar) {
                constraintValue = TimeUtil.formatCalendar((Calendar) value);
            } else {
                constraintValue = value.toString();
                if (value.getClass().isArray() || value instanceof Collection) {
                    constraintType = ConstraintType.IN;
                }
            }
            RequestConstraint constraint = new RequestConstraint(
                    constraintValue, constraintType);
            if (constraintType == ConstraintType.IN) {
                // Set constraint value list
                List<String> constraintValueList = new ArrayList<String>();
                if (value.getClass().isArray()) {
                    int size = Array.getLength(value);
                    for (int i = 0; i < size; ++i) {
                        Object arrayValue = Array.get(value, i);
                        if (arrayValue != null) {
                            constraintValueList.add(arrayValue.toString());
                        }
                    }
                } else if (value instanceof Collection) {
                    for (Object collValue : ((Collection<?>) value)) {
                        if (collValue != null) {
                            constraintValueList.add(collValue.toString());
                        }
                    }
                }
                constraint.setConstraintValueList(constraintValueList);
            }
            if (value != null || includeNulls) {
                constraints.put(key, constraint);
            }
        }
        return constraints;
    }
}

File: pythonPackages/dynamicserialize/dstypes/com/raytheon/uf/common/dataquery/requests/RequestConstraint.py
##
# This software was developed and / or modified by Raytheon Company,
# pursuant to Contract DG133W-05-CQ-1067 with the US Government.
#
# U.S. EXPORT CONTROLLED TECHNICAL DATA
# This software product contains export-restricted data whose
# export/transfer/disclosure is restricted by U.S. law. Dissemination
# to non-U.S. persons whether in the United States or abroad requires
# an export license or other authorization.
#
# Contractor Name:        Raytheon Company
# Contractor Address:     6825 Pine Street, Suite 340
#                         Mail Stop B8
#                         Omaha, NE 68106
#                         402.291.0100
#
# See the AWIPS II Master Rights File ("Master Rights File.pdf") for
# further licensing information.
##

#
#      SOFTWARE HISTORY
#
#     Date            Ticket#       Engineer       Description
#     ------------    ----------    -----------    --------------------------
#     Jun 01, 2016    5574          tgurney        Initial creation
#     Jun 27, 2016    5725          tgurney        Add NOT IN
#     Jul 22, 2016    2416          tgurney        Add evaluate()
#
#

import re
from ...time import DataTime


class RequestConstraint(object):

    TOLERANCE = 0.0001

    IN_PATTERN = re.compile(',\s?')

    def __init__(self):
        self.constraintValue = None
        self.constraintType = None

    def getConstraintValue(self):
        return self.constraintValue

    def setConstraintValue(self, constraintValue):
        if hasattr(self, '_evalValue'):
            del self._evalValue
        self.constraintValue = constraintValue

    def getConstraintType(self):
        return self.constraintType

    def setConstraintType(self, constraintType):
        if hasattr(self, '_evalValue'):
            del self._evalValue
        self.constraintType = constraintType

    def evaluate(self, value):
        if not hasattr(self, '_evalValue'):
            self._setupEvalValue()

        if self.constraintType == 'EQUALS':
            return self._evalEquals(value)
        elif self.constraintType == 'NOT_EQUALS':
            return not self._evalEquals(value)
        elif self.constraintType == 'GREATER_THAN':
            return self._evalGreaterThan(value)
        elif self.constraintType == 'GREATER_THAN_EQUALS':
            return self._evalGreaterThanEquals(value)
        elif self.constraintType == 'LESS_THAN':
            return self._evalLessThan(value)
        elif self.constraintType == 'LESS_THAN_EQUALS':
            return self._evalLessThanEquals(value)
        elif self.constraintType == 'BETWEEN':
            return self._evalBetween(value)
        elif self.constraintType == 'IN':
            return self._evalIn(value)
        elif self.constraintType == 'NOT_IN':
            return not self._evalIn(value)
        elif self.constraintType == 'LIKE':
            return self._evalLike(value)
        # setupConstraintType already adds correct flags for ilike
        # on regex pattern
        elif self.constraintType == 'ILIKE':
            return self._evalLike(value)
        elif self.constraintType == 'ISNULL':
            return self._evalIsNull(value)
        elif self.constraintType == 'ISNOTNULL':
            return not self._evalIsNull(value)
        else:
            errmsg = '{} is not a valid constraint type.'
            raise ValueError(errmsg.format(self.constraintType))

    def _makeRegex(self, pattern, flags):
        """Make a pattern using % wildcard into a regex"""
        pattern = re.escape(pattern)
        pattern = pattern.replace('\\%', '.*')
        pattern = pattern.replace('\\_', '.')
        pattern = pattern + '$'
        return re.compile(pattern, flags)

    def _setupEvalValue(self):
        if self.constraintType == 'BETWEEN':
            self._evalValue = self.constraintValue.split('--')
            self._evalValue[0] = self._adjustValueType(self._evalValue[0])
            self._evalValue[1] = self._adjustValueType(self._evalValue[1])
        elif self.constraintType in ('IN', 'NOT_IN'):
            splitValue = self.IN_PATTERN.split(self.constraintValue)
            self._evalValue = {
                self._adjustValueType(value)
                for value in splitValue
                }
            # if collection now contains multiple types we have to force
            # everything to string instead
            initialType = next(iter(self._evalValue)).__class__
            for item in self._evalValue:
                if item.__class__ is not initialType:
                    self._evalValue = {str(value) for value in splitValue}
                    break
        elif self.constraintType == 'LIKE':
            self._evalValue = self._makeRegex(self.constraintValue, re.DOTALL)
        elif self.constraintType == 'ILIKE':
            self._evalValue = self._makeRegex(self.constraintValue, re.IGNORECASE | re.DOTALL)
        elif self.constraintValue is None:
            self._evalValue = None
        else:
            self._evalValue = self._adjustValueType(self.constraintValue)

    def _adjustValueType(self, value):
        '''
        Try to take part of a constraint value, encoded as a string, and
        return it as its 'true type'.

        _adjustValueType('3.0') -> 3.0
        _adjustValueType('3') -> 3.0
        _adjustValueType('a string') -> 'a string'
        '''
        try:
            return float(value)
        except Exception:
            pass
        try:
            return DataTime(value)
        except Exception:
            pass
        return value

    def _matchType(self, value, otherValue):
        '''
        Return value coerced to be the same type as otherValue. If this is
        not possible, just return value unmodified.
        '''
        # cannot use type() because otherValue might be an instance of an
        # old-style class (then it would just be of type "instance")
        if not isinstance(value, otherValue.__class__):
            try:
                return otherValue.__class__(value)
            except Exception:
                pass
        return value

    def _evalEquals(self, value):
        value = self._matchType(value, self._evalValue)
        if isinstance(value, float):
            return abs(float(self._evalValue) - value) < self.TOLERANCE
        else:
            return value == self._evalValue

    def _evalGreaterThan(self, value):
        value = self._matchType(value, self._evalValue)
        return value > self._evalValue

    def _evalGreaterThanEquals(self, value):
        value = self._matchType(value, self._evalValue)
        return value >= self._evalValue

    def _evalLessThan(self, value):
        value = self._matchType(value, self._evalValue)
        return value < self._evalValue

    def _evalLessThanEquals(self, value):
        value = self._matchType(value, self._evalValue)
        return value <= self._evalValue

    def _evalBetween(self, value):
        value = self._matchType(value, self._evalValue[0])
        return value >= self._evalValue[0] and value <= self._evalValue[1]

    def _evalIn(self, value):
        anEvalValue = next(iter(self._evalValue))
        if isinstance(anEvalValue, float):
            for otherValue in self._evalValue:
                try:
                    if abs(otherValue - float(value)) < self.TOLERANCE:
                        return True
                except Exception:
                    pass
            return False
        else:
            value = self._matchType(value, anEvalValue)
            return value in self._evalValue

    def _evalLike(self, value):
        value = self._matchType(value, self._evalValue)
        if self.constraintValue == '%':
            return True
        return self._evalValue.match(value) is not None

    def _evalIsNull(self, value):
        return value is None or 'null' == value

    # DAF-specific stuff begins here ##########################################

    CONSTRAINT_MAP = {'=': 'EQUALS',
                      '!=': 'NOT_EQUALS',
                      '>': 'GREATER_THAN',
                      '>=': 'GREATER_THAN_EQUALS',
                      '<': 'LESS_THAN',
                      '<=': 'LESS_THAN_EQUALS',
                      'IN': 'IN',
                      'NOT IN': 'NOT_IN'
                      }

    @staticmethod
    def _stringify(value):
        if type(value) in {str, int, long, bool, float, unicode}:
            return str(value)
        else:
            # Collections are not allowed; they are handled separately.
            # Arbitrary objects are not allowed because the string
            # representation may not be sufficient to reconstruct the object.
            raise TypeError('Constraint values of type ' + repr(type(value)) +
                            'are not allowed')

    @classmethod
    def _constructIn(cls, constraintType, constraintValue):
        """Build a new "IN" or "NOT IN" constraint from an iterable."""
        try:
            iterator = iter(constraintValue)
        except TypeError:
            raise TypeError("value for IN / NOT IN constraint must be an iterable")
        stringValue = ', '.join(cls._stringify(item) for item in iterator)
        if len(stringValue) == 0:
            raise ValueError('cannot use IN / NOT IN with empty collection')
        obj = cls()
        obj.setConstraintType(constraintType)
        obj.setConstraintValue(stringValue)
        return obj

    @classmethod
    def _constructEq(cls, constraintType, constraintValue):
        """Build a new = or != constraint. Handle None specially by making an
        "is null" or "is not null" instead.
        """
        obj = cls()
        if constraintValue is None:
            if constraintType == 'EQUALS':
                obj.setConstraintType('ISNULL')
            elif constraintType == 'NOT_EQUALS':
                obj.setConstraintType('ISNOTNULL')
        else:
            obj = cls._construct(constraintType, constraintValue)
        return obj

    @classmethod
    def _construct(cls, constraintType, constraintValue):
        """Build a new constraint."""
        stringValue = cls._stringify(constraintValue)
        obj = cls()
        obj.setConstraintType(constraintType)
        obj.setConstraintValue(stringValue)
        return obj

    @classmethod
    def new(cls, operator, constraintValue):
        """Build a new RequestConstraint."""
        try:
            constraintType = cls.CONSTRAINT_MAP[operator.upper()]
        except KeyError, AttributeError:
            errmsg = '{} is not a valid operator. Valid operators are: {}'
            validOperators = list(sorted(cls.CONSTRAINT_MAP.keys()))
            raise ValueError(errmsg.format(operator, validOperators))
        if constraintType in ('IN', 'NOT_IN'):
            return cls._constructIn(constraintType, constraintValue)
        elif constraintType in {'EQUALS', 'NOT_EQUALS'}:
            return cls._constructEq(constraintType, constraintValue)
        else:
            return cls._construct(constraintType, constraintValue)

